### ISSUE TITLE
Warning: setState(...): Cannot update during an existing state transition

### ISSUE THREAD

**Issue by ObedAmpah** [react-tools/react-form#202]
Hello,

I receive a `warning` with my custom component, which is an input I need to display a calculated sum.

`Warning: setState(...): Cannot update during an existing state transition (such as within `render` or another component's constructor). Render methods should be a pure function of props and state; constructor side-effects are an anti-pattern, but can be moved to `componentWillMount`.
`
I am doing the calculation in the parent using state to return the value of a reduced object of values, 
```js
this.state[key]
```
Below is my custom input component:
```html
<CustomInput value={this.state[key]} field={...} />
```

Below is my input:          
```html
          <input
                {...rest}
                readOnly={true}
                value={getValue() || ' '}
                onChange={setValue()}
            />
```

Below is the logic for taking in objects made up of columns of form data and then reducing them into per/key sum values. This runs in the parent component.
```js
            // Loop over the values found in the values array
            return formDataObj.map((item, key) => {

                // I update the state here to show the calculation change for each new added value in my table
                this.setState({ [key+1]: sumValues(formDataObjLoop[key]) })


                // Add to the formData data object
                formApi.setValue("Column_Total_" + [key+1], sumValues(formDataObjLoop[key]))
            })
```

My situation, is that the value i'm passing down into the component is given by state, because I am not touching the input and updating the value as it would normally expect.

Thoughts on what I could be doing wrong? 

Thanks

**Comment by joepuzzo** [comment]
You should never call set state in a render. It can cause loops and other bad things. You have two issues in your code. 1: you are calling setState in a render. 2: you are calling setValue in a render. Move this logic up and out of the render. If you want to setValues then you should use the getApi or defaultValues props that are provided, see docs for details

**Comment by ObedAmpah** [comment]
@joepuzzo thanks for reaching out! 

I know that I can calculate the sum of values and then pass it to my custom input.

Can I, then,  pass that value into `getValue()` and have it taken up by the example get-set logic in the custom input example?

I noticed I can't get an `onChange={(e) => { this.handleChange(e) }}` event to trigger with the value updating.

Please excuse any no0b-ness to my question, just trying to get a breakthrough so I can move forward.

Thanks,

**Comment by joepuzzo** [comment]
My understanding is that you want to set a value in a field dynamically. For example, maybe when you update the number in field A, you want the number in field B to get updated based on that change and some math. To do this you simply need to 1. Hook into the on change event of field A and have that call a funtion that calculates your new value. Then 2. That same function will then call setValue. But remember, this function should live outside of render, and it will have access to setValue because you will save the formApi to a member variable or somthing.

**Comment by joepuzzo** [comment]
Can we close this ?

**Comment by ObedAmpah** [comment]
Yes! I was able to get the value set and save it to the values object. My
concern was getting the value into the nested object I created using the
field syntax for my dynamically generated form inputs.

Field Syntax was great for creating groupings for columns in my table. I
may resort to passing a UID down the columns of my table to reduce the
values for my totals now so the values object isn't nested (I don't have
any array vals I need to capture in this form)

Thanks for your help with my earlier issue.
